### ISSUE TITLE
node upgrade failed with 'node not found' in v2.6.4

### ISSUE THREAD

**Issue by Shuanglu** [rancher/rancher#37634]
**Rancher Server Setup**
- Rancher version: v2.6.4
- Installation option (Docker install/Helm Chart):
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc): helm/v1.18.18 provisioned by rke
- Proxy/Cert Details:N/A

**Information about the Cluster**
- Kubernetes version: v1.20.11
- Cluster Type (Local/Downstream):
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider): Custom Downstream running on EC2

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
  - If custom, define the set of permissions: Admin
   

   
**Describe the bug**

We tried to operate the cluster via terraform and it got stuck with error below. 
```
[controlPlane] Failed to upgrade Control Plane: [[[controlplane] Error getting node ip-xxxx-us-east-2.compute.internal: "ip-xxx-us-east-2.compute.internal" not found]]
```
The node actually exists and it turns out to be the 'kubernetes.io/hostname' is the short hostname but the rancher validates the node existence according to the full hostname. Prior to v2.6.4 the parameter '--node-name "$(hostname -f)"' we added to the node registration command worked fine. However in rancher v2.6.4, it doesn't work any more and returned such error. We have to manually patch the node label to unblock the operation. 
We found out there was one commit which deleted the parameter 'hostname-override' in rke 1.3.8 https://github.com/rancher/rke/pull/2803

**To Reproduce**
 
- install rancher v2.6.4 
- provision a downstream cluster and add '--node-name' to overwrite the short hostname 

**Result**

the label 'kubernetes.io/hostname' of the registered node is the short one instead of the full hostname. rancher may report error 'node not found'

**Expected Result**

rancher handle the inconsistence of the values smoothly. If rancher generate the 'node.status.hostnameOverride' according to the "RequestedHostName", it shouldn't remove the hostname user specified when generate the nodeplan. Else rancher shall bypass the same to validate. 
I'm not sure what's the story behind this design and found it might be related with below code but feel free to correct me 
nodecreation: https://github.com/rancher/rancher/blob/v2.6.4/pkg/controllers/management/node/controller.go#L124 
nodeplan: https://github.com/rancher/rke/blob/v1.3.8/cluster/plan.go#L477 
validation:  https://github.com/rancher/rke/blob/v1.3.8/k8s/node.go#L55
**Screenshots**

**Additional context**

SURE-4375
SURE-4912

**Comment by snasovich** [comment]
Setting the milestone to v2.6.6 to increase the visibility and add it to our team's queue; not a commitment to deliver the fix in 2.6.6.

**Comment by Shuanglu** [comment]
@kinarashah just wonder do we have plan of the fix or ETA

**Comment by bkonick** [comment]
I believe that this is similar to the issue I opened here https://github.com/rancher/rke/issues/3091

**Comment by kinarashah** [comment]
**Kubelet**
---- 
**To reproduce**

Rancher fresh install >= v2.6.4 
1. Custom clusters
- Custom cluster (set --node-name to a value that doesn't match with the default hostname assigned for eg "hostname -f" or "kinara-ec2")
- Cluster provisioning fails with node xxxx not found  error. 

2. Node Driver clusters
- Cluster provisioning succeeds. This is because rancher defaults to the private ip address of the node correctly which is what shows up as the kubernetes.io/hostname label.

3. Rancher upgrade path 
- 2.6.3: Custom cluster (set --node-name to a value that doesn't match with the default hostname assigned for eg "hostname -f" or "kinara-ec2")
- Upgrade to 2.6.4
- Upgrade cluster's k8s version or change rkeConfig to trigger a cluster upgrade.
- Cluster provisioning fails with node not found  error.

4. RKE standalone 
- Set hostname_override as mentioned above or do not set hostname_override - RKE defaults it to the public ip address of the node. 
- Cluster provisioning fails with node not found  error. 

**Workarounds:** 
- Edit the [kubernetes.io/hostname](http://kubernetes.io/hostname) label on the node with xxxx from node xxxx not found error. 
- Change the hostname of the node to the name cloud provider generates, this is usually the private DNS of the node returned by the ec2 metadata service. (curl http://169.254.169.254/latest/meta-data/local-hostname) but this is not guaranteed when using custom dns or when hostname is set differently and is not the same as the one returned by ec2 metadata.

**To Test:** 

Fresh Installs
Tests 1-4 pass with fresh install of Rancher v2.6-head and RKE v1.3.x-RC 

Upgrades 
- Tests 1-4 reproduced with Rancher v2.6.4 
- Upgrade to Rancher v2.6-head 
- Tests 1-4 pass with Rancher v2.6-head

Upgrade path after implementing workarounds 
- Tests 1-4 reproduced with Rancher v2.6.4 
- Implement workaround 1 and 2 on different clusters 
- Upgrade to Rancher v2.6-head 
- Tests 1-4 pass with Rancher v2.6-head

**Kube-proxy**
---- 
- Issues https://github.com/rancher/rancher/issues/22416 and https://github.com/rancher/rancher/issues/18483 continue to work before and after the fix. 
- Test usecases around `useInstanceMetadataHostname`: should be default false, no nil panics and turning on/off should reset the kube-proxy args for `hostname-override`.

**Comment by kinarashah** [comment]
Available to test with RKE `v1.3.18-rc3` and Rancher v2.6-head once https://drone-publish.rancher.io/rancher/rancher/8841 is green.

**Comment by Josh-Diamond** [comment]
Ticket #37634 - Test Plan
---

**Kubelet:** 
- Run the following scenarios for each: - [`cloud provider enabled`]
    - Custom cluster
    - Node driver Cluster
    - Rancher upgrade path
    - RKE Standalone
    
| Scenario | Test Case | Expected Result |
|---|---|---|
| 1. | Fresh install of rancher `v2.6-head` w/ RKE `v1.3.18-rc3` + follow reproduction steps above | Clusters should successfully provision: With clean logs if Node Name matches cloud provider assigned hostname, and with logs if Node Name does NOT match cloud provider assigned hostname, logs will inform you hostname has been re-assigned to match cloud provider assigned hostname  |
| 2. | Reproduce in `v2.6.4` => upgrade to `v2.6-head` | Clusters should reproduce issue in `v2.6.4`; clusters should not reproduce issue in `v2.6-head` |
| 3. | Reproduce in `v2.6.4` => implement workarounds => upgrade to `v2.6-head` | Clusters should reproduce issue in `v2.6.4`; workarounds should resolve initial error; clusters should not reproduce issue in `v2.6-head` |

---

**Kube-proxy**
| Scenario | Test Case | Result |
|---|---|---|
| 1. | Provision Node driver cluster on `2.6-head` | Cluster should successfully provision; `useInstanceMetadataHostname` should default to `true` for Node driver clusters |
| 2. | Provision Node driver cluster on `2.6-head`, set `useInstanceMetadataHostname` to `false` + misconfigure Node name from cloud provider assigned hostname | Cluster should fail to provision |
| 3. | Provision Node driver cluster on `2.6-head`, set `useInstanceMetadataHostname` to `true` + misconfigure Node name from cloud provider assigned hostname | Cluster should successfully provision with logs detailing the Node name was overwritten by Cloud provider assigned hostname |
| 4. | Provision Custom cluster on `2.6-head` + misconfigure Node name from cloud provider assigned hostname | `useInstanceMetadataHostname` should default to `false` for Custom clusters; cluster is expected to fail |
| 5. | Provision Custom cluster on `2.6-head` + accurately configure Node name to cloud provider assigned hostname | `useInstanceMetadataHostname` should default to `false` for Custom clusters; cluster is expected to successfully provision |
| 6. | Provision Custom cluster on `2.6-head` + set `useInstanceMetadataHostname` to `true` | cluster may or may not provision, depending on the custom cluster configurations but it is expected that the Node name will be set to the assigned Cloud provider hostname |
| 7. | Provision RKE standalone cluster w/ misconfigured Node name from cloud provider assigned hostname | `useInstanceMetadataHostname` should default to `false`; cluster expected to fail provisioning |
| 8. | Provision RKE standalone cluster + configured Node name to match cloud provider assigned hostname | `useInstanceMetadataHostname` should default to `false`; cluster expected to successfully provision |

**Comment by sowmyav27** [comment]
@rishabhmsra Could you please reverify this usecase - https://github.com/rancher/rancher/issues/22416 on 2.6-head with changes?

**Comment by Josh-Diamond** [comment]
Ticket #37634 - Reproduced - ✅ 
---

With Docker on a single-node instance:

---

Scenario 1 - Fresh install on rancher `2.6.4`:
1. Fresh install of Rancher `v2.6.4`
2. As admin, create a standard user
3. As standard user, Provision a downstream RKE1 Custom cluster w/ AWS Cloud provider enabled, but misconfigure `--node-name` in registration command to anything other than the actual node names
4. Reproduced:
 ```logs
 [controlPlane] Failed to upgrade Control Plane: [[[controlplane] Error getting node <REDACTED>: "<REDACTED>" not found]]
 ```
 
 Screenshot:
<img width="1599" alt="node-not-found-reproduced" src="https://user-images.githubusercontent.com/46494969/213801942-b1ac928e-8fbd-4a78-ab6e-56f7972f30cc.png">

 ---
 
 Scenario 2 - Upgrade - [ `v2.6.3` => `v2.6.4`]:
1. Fresh install of Rancher `v2.6.3`
2. As admin, create a standard user
3. As standard user, Provision a downstream RKE1 Custom cluster w/ AWS Cloud provider enabled, but misconfigure `--node-name` in registration command to anything other than the actual node names
4. Observe cluster successfully provision
5. Upgrade Rancher to `v2.6.4`
6. Once Rancher has successfully upgraded, `Edit Config` of cluster and upgrade k8s version
7. Reproduced:
 ```logs
 [controlPlane] Failed to upgrade Control Plane: [[[controlplane] Error getting node <REDACTED>: "<REDACTED>" not found]]
 ```

Screenshots:

`Step 4`
<img width="1368" alt="Screenshot 2023-01-20 at 12 44 26 PM" src="https://user-images.githubusercontent.com/46494969/213802888-faaba570-2d0b-4af3-9e03-b4daf1108c27.png">

`Step 7`
<img width="1361" alt="Screenshot 2023-01-20 at 1 09 25 PM" src="https://user-images.githubusercontent.com/46494969/213802947-0503f958-ebeb-4608-9244-714a534e7cb2.png">

**Comment by Josh-Diamond** [comment]
Ticket #37634 - Test Results - ✅ 
---

With Docker on a single-node instance:

Verified on rancher `v2.6-b465433733eaec845bcfd8b67f15ef5d33e659e8-head`:

**Kubelet**

| Scenario | Test Results | Status |
|---|---|---|
| 1. Fresh install [`2.6-head`] | Provision a downstream RKE1 Custom cluster w/ Amazon Cloud provider enabled + Node name != hostname | ✅ |
| 2. Fresh install - [`Node driver; regression check`] | Provision a downstream RKE1 Node driver cluster w/ Amazon Cloud provider enabled | ✅ |
| 3. Upgrade `2.6.3 => 2.6-head` | Provision a downstream RKE1 Custom cluster w/ Amazon Cloud provider enabled + Node name != hostname on `v2.6.3` => upgrade to `v2.6-head` => add additional worker node | ✅ |
| 4. Upgrade Path | Provision a downstream RKE1 Custom cluster w/ Amazon Cloud provider enabled + Node name != hostname on `v2.6.3` => upgrade to `v2.6.4` + update k8s version to reproduce => upgrade rancher to `v2.6-head` + upgrade k8s version | ✅ |
| 5. Upgrade Path [`workaround #1`] | Provision a downstream RKE1 Custom cluster w/ Amazon Cloud provider enabled + Node name != hostname on `v2.6.3` => upgrade to `v2.6.4` + update k8s version to reproduce => edit `kubernetes.io/hostname` label + upgrade k8s version to resolve error => upgrade rancher to `v2.6-head` + add worker node | ✅ |
| 6. Upgrade Path [`workaround #2`] |  Provision a downstream RKE1 Custom cluster w/ Amazon Cloud provider enabled + Node name != hostname on `v2.6.3` => upgrade to `v2.6.4` + update k8s version to reproduce => edit node name via AWS Console + upgrade k8s version to resolve error => upgrade rancher to `v2.6-head` + add worker node | ✅ |

---

**Kube-proxy***

| Scenario | Test Results | Status |
|---|---|---|
| 1. Fresh install | Provision Custom cluster on 2.6-head + misconfigure Node name from cloud provider assigned hostname => Verify hostname-override value to args of kube-proxy value is same as node-name passed by registration cmd | ✅ |
| 2. Fresh Install - [Custom; `Node name == hostname`] | Provision Custom cluster on 2.6-head + configure Node name to match cloud provider assigned hostname | ✅ |
| 3. Fresh Install - [Custom; `useInstanceMetadataHostname = true`] | Provision Custom cluster on 2.6-head + set `useInstanceMetadataHostname = true` => Verify hostname-override value to args of kube-proxy value is same Instance's metadata hostname | ✅ |
| 4. Upgrade - [`Custom`] | Provision custom cloud provider enabled AWS Cluster on `v2.6.9` => Upgrade rancher to `v2.7-head` => Add a new worker node to the cluster | ✅ |

---

Full testing report available internally via Confluence

Note: `useInstanceMetadataHostname` is not present on cluster.yml after provisioning by default; an additional ticket will be created to further track and address this

**Comment by sowmyav27** [comment]
Reopening this for a few additional checks

**Comment by rishabhmsra** [comment]
Validated the usecase as mentioned here: https://github.com/rancher/rancher/issues/37634#issuecomment-1397985310

- **On rancher v2.7-head(9376ad2)**
  - Created following single node all role, ec2 node driver clusters with Amazon(in-tree) enabled - v1.25.6-rancher1-1, v1.24.10-rancher3-1, v1.23.16-rancher1-1. Also while provisioning, make sure to set `useInstanceMetadataHostname` as `true`
  
```
cloud_provider:
  name: aws
  useInstanceMetadataHostname: true
```
   - SSH'ed into cluster nodes.
   - Validated `--hostname-override` arg is present on the `kube-proxy` process and with value set as the hostname provided by the cloud provider.
------------------------------------
- **On rancher v2.6-head(4db94f6)**
  - Created following single node all role, ec2 node driver clusters with Amazon(in-tree) enabled - v1.24.9-rancher1-1, v1.23.15-rancher1-1. Also while provisioning, make sure to set `useInstanceMetadataHostname` as `true`
 
```
cloud_provider:
  name: aws
  useInstanceMetadataHostname: true
```
  - SSH'ed into cluster nodes.
  - Validated `--hostname-override` arg is present on the `kube-proxy` process and with value set as the hostname provided by the cloud provider.

**Result :**
```
ps -ef | grep -i kube-proxy
root       45153   45131  0 07:32 ?        00:00:02 kube-proxy --kubeconfig=/etc/kubernetes/ssl/kubecfg-kube-proxy.yaml --healthz-bind-address=127.0.0.1 --v=2 --bind-address=<redacted> --cluster-cidr=<redacted>/16 --hostname-override=ip-<redacted>.us-east-2.compute.internal
```
- Also created custom ingress + NLB as mentioned [here](https://github.com/rancher/rancher/issues/22416#issue-485187046) on all downstream clusters. LoadBalancer service created for ingress-nginx-controller and on accessing the `http://<EXTERNAL-IP>/healthz` getting `HTTP/1.1 200 OK` response

--------------------------------

**Comment by kinarashah** [comment]
@sowmyav27 

The issue is NodePort service stops working when patched with `'externalTrafficPolicy=Local'`. This happens when `kube-proxy` doesn't have the same hostname as the node name (the name k8s registered the node as). https://github.com/rancher/rancher/issues/18483#issuecomment-511188390 

So to test this after the fix, we should confirm the following:
1. Without `useInstanceMetadataHostname` - 
rke sets--hostname-override to the hostname override set by the user. `externalTrafficPolicy=Local` doesn't work.
2. enable/set `useInstanceMetadataHostname` to true - 
rke sets --hostname-override to the value provided by the cloud provider and thus matches with the name k8s registers the node. `externalTrafficPolicy=Local` works.

**Comment by rishabhmsra** [comment]
Validated on rancher v2.6-head(657d8ec)

**Scenario 1 :** without `useInstanceMetadataHostname`

- Created a k8s v1.24.10-rancher3-1 cluster - single node, all role with Amazon(in-tree) enabled.
- `--hostname-override` is NOT set to the node name provided by cloud provider:
```
ps -ef | grep -i kube-proxy
root        9390    9367  0 06:58 ?        00:00:00 kube-proxy --cluster-cidr=<REDACTED>/16 --hostname-override=rishabh-sn1 --kubeconfig=/etc/kubernetes/ssl/kubecfg-kube-proxy.yaml --healthz-bind-address=127.0.0.1 --v=2 --bind-address=<REDACTED>
```
- Created deployment, service and patched the service with "externalTrafficPolicy":"Local" as mentioned [here](https://github.com/rancher/rancher/issues/18483#issue-415771391)
```
kubectl describe svc nodeport | grep External
External Traffic Policy:  Local
```
- nodeport service stopped responding. Got NO response from `for node in $NODES; do curl -s $node:$NODEPORT | grep -i client_address; done`

--------------------------

**Scenario 2 :** with enable/set `useInstanceMetadataHostname` to `true`
- Created a k8s v1.24.10-rancher3-1 cluster - single node, all role with Amazon(in-tree) enabled.
- `--hostname-override` is set to the node name provided by cloud provider:
```
root        9438    9414  0 07:02 ?        00:00:00 kube-proxy --cluster-cidr=<REDACTED>/16 --kubeconfig=/etc/kubernetes/ssl/kubecfg-kube-proxy.yaml --healthz-bind-address=127.0.0.1 --v=2 --bind-address=<REDACTED> --hostname-override=ip-<REDACTED>.us-east-2.compute.internal
```
- Created deployment, service and patched the service with "externalTrafficPolicy":"Local" as mentioned [here](https://github.com/rancher/rancher/issues/18483#issue-415771391)
```
kubectl describe svc nodeport | grep External
External Traffic Policy:  Local
```
- externalTrafficPolicy=Local works:
```
for node in $NODES; do curl -s $node:$NODEPORT | grep -i client_address; done
client_address=<REDACTED>
```

**Comment by rishabhmsra** [comment]
Validated on rancher v2.6-head(657d8ec)

**Scenario 1 :** without `useInstanceMetadataHostname`

- Created a k8s v1.24.10-rancher3-1 cluster - single node, all role with Amazon(in-tree) enabled.
- `--hostname-override` is NOT set to the node name provided by cloud provider:
```
ps -ef | grep -i kube-proxy
root        9390    9367  0 06:58 ?        00:00:00 kube-proxy --cluster-cidr=<REDACTED>/16 --hostname-override=rishabh-sn1 --kubeconfig=/etc/kubernetes/ssl/kubecfg-kube-proxy.yaml --healthz-bind-address=127.0.0.1 --v=2 --bind-address=<REDACTED>
```
- Created deployment, service and patched the service with "externalTrafficPolicy":"Local" as mentioned [here](https://github.com/rancher/rancher/issues/18483#issue-415771391)
```
kubectl describe svc nodeport | grep External
External Traffic Policy:  Local
```
- nodeport service stopped responding. Got NO response from `for node in $NODES; do curl -s $node:$NODEPORT | grep -i client_address; done`

--------------------------

**Scenario 2 :** with enable/set `useInstanceMetadataHostname` to `true`
- Created a k8s v1.24.10-rancher3-1 cluster - single node, all role with Amazon(in-tree) enabled.
- `--hostname-override` is set to the node name provided by cloud provider:
```
root        9438    9414  0 07:02 ?        00:00:00 kube-proxy --cluster-cidr=<REDACTED>/16 --kubeconfig=/etc/kubernetes/ssl/kubecfg-kube-proxy.yaml --healthz-bind-address=127.0.0.1 --v=2 --bind-address=<REDACTED> --hostname-override=ip-<REDACTED>.us-east-2.compute.internal
```
- Created deployment, service and patched the service with "externalTrafficPolicy":"Local" as mentioned [here](https://github.com/rancher/rancher/issues/18483#issue-415771391)
```
kubectl describe svc nodeport | grep External
External Traffic Policy:  Local
```
- externalTrafficPolicy=Local works:
```
for node in $NODES; do curl -s $node:$NODEPORT | grep -i client_address; done
client_address=<REDACTED>
```

**Comment by rishabhmsra** [comment]
Above scenarios are working as expected, hence closing the issue.

**Comment by kinarashah** [comment]
Placeholder comment for summarizing changes (linked in release notes for Rancher v2.6.11)
